### PR TITLE
Fast HotStuff Vote Msg Storage

### PR DESCRIPTION
**Pull request by tholonious**

### CODE REVIEW

**Comment by tholonious** [comment]
Current dependencies on/for this PR:
* feature/pos-fast-hotstuff-consensus
  * **PR #640** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/640" target="_blank">Graphite</a> 
    * **PR #641** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/641" target="_blank">Graphite</a> 
      * **PR #642** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/642" target="_blank">Graphite</a> 
        * **PR #643** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/643" target="_blank">Graphite</a> 
          * **PR #645** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/645" target="_blank">Graphite</a> 
            * **PR #648** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/648" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/deso-protocol/core/648?utm_source=stack-comment).

File: consensus/consensus.go [hunk 1]
```diff
@@ -164,8 +167,75 @@ func (fc *FastHotStuffConsensus) AdvanceView() (uint64, error) {
...
+		return errors.Errorf(
+			"FastHotStuffConsensus.CaptureValidatorVote: validator %s has already timed out for view %d",
+			vote.GetPublicKey().ToString(),
+			vote.GetView(),
+		)
+	}
+
+	// Note: we do not check if the vote is for the current chain tip's blockhash. During leader changes
+	// where we will be the next block proposer, it is possible for us to receive a vote for a block that
+	// we haven't seen yet, but we will need to construct the a QC for the block as we are the next leader.
```
Comment (lazynina) [inline consensus/consensus.go:232]:
```suggestion
	// we haven't seen yet, but we will need to construct the QC for the block as we are the next leader.
```

File: consensus/consensus.go [hunk 2]
```diff
@@ -306,3 +376,49 @@ func (fc *FastHotStuffConsensus) evictStaleVotesAndTimeouts() {
...
+	return false
+}
+
+func (fc *FastHotStuffConsensus) hasTimedOutForView(publicKey *bls.PublicKey, view uint64) bool {
+	timeoutsForView, ok := fc.timeoutsSeen[view]
+	if !ok {
+		return false
+	}
+
+	_, ok = timeoutsForView[publicKey.ToString()]
```
Comment (lazynina) [inline consensus/consensus.go:418]: The mere existence of a public key in this map indicates we have a valid timeout message, correct?

File: consensus/consensus_test.go [hunk 3]
```diff
@@ -256,6 +256,82 @@ func TestAdvanceView(t *testing.T) {
...
+	{
+		currentView, err := fc.AdvanceView()
+		require.NoError(t, err)
+		require.Equal(t, uint64(3), currentView)
+	}
+
+	// Test with malformed vote
+	{
+		err := fc.CaptureValidatorVote(nil)
+		require.Error(t, err)
```
Comment (lazynina) [inline consensus/consensus_test.go:281]: Do we want to check for specific errors in this test?

File: consensus/utils.go [hunk 4]
```diff
@@ -1,17 +1,34 @@
...
 import (
+	"crypto/sha256"
+	"fmt"
 	"reflect"
 
+	"github.com/deso-protocol/core/bls"
 	"github.com/deso-protocol/core/collections"
 )
 
+// When voting on a block, validator sign the payload sha256(View, BlockHash) with their BLS
```
Comment (lazynina) [inline consensus/utils.go:12]:
```suggestion
// When voting on a block, validators sign the payload sha256(View, BlockHash) with their BLS
```

File: consensus/utils_test.go [hunk 5]
```diff
@@ -84,6 +84,40 @@ func TestIsProperlyFormedValidatorSet(t *testing.T) {
...
+		vote.publicKey = nil
+		require.False(t, isProperlyFormedVote(vote))
+	}
+
+	// Test nil signature
+	{
+		vote := createDummyVoteMessage(1)
+		vote.signature = nil
+		require.False(t, isProperlyFormedVote(vote))
+	}
```
Comment (lazynina) [inline consensus/utils_test.go:118]: Happy path test?

**Review by lazynina — CHANGES_REQUESTED** [review]
Just a couple of nits

File: consensus/consensus.go (hunk 2 above)
Comment (tholonious) [reply consensus/consensus.go:418]: Correct

File: consensus/consensus_test.go (hunk 3 above)
Comment (tholonious) [reply consensus/consensus_test.go:281]: Done

**Review by lazynina — APPROVED** [review]

**Review by AeonSw4n — CHANGES_REQUESTED** [review]

File: consensus/event_loop.go [hunk 6]
```diff
@@ -163,8 +166,75 @@ func (fc *FastHotStuffEventLoop) ProcessSafeBlock(block Block, validators []Vali
...
+	// Do a basic integrity check on the vote message
+	if !isProperlyFormedVote(vote) {
+		return errors.New("FastHotStuffEventLoop.ProcessValidatorVote: Malformed vote message")
+	}
+
+	// Compute the value sha256(vote.View, vote.BlockHash)
+	voteSignaturePayload := GetVoteSignaturePayload(vote.GetView(), vote.GetBlockHash())
+
+	// Verify the vote signature
+	if !isValidSignature(vote.GetPublicKey(), vote.GetSignature(), voteSignaturePayload[:]) {
```
[the remark this answers is not in the document]
Comment (tholonious) [reply consensus/event_loop.go:200]: Good catch. I've moved the signature verification to the end

**Review by AeonSw4n — APPROVED** [review]

**Pull request merged**